### PR TITLE
Fix for avatar opening executable

### PR DESCRIPTION
**Pull request by nandesu-utils**
In issue #1742 the reason behind crashing is supposedly undefined behavior because it was observed by me, `g_spawn_sync` may not modify `spawn_error` pointer. When I was debugging, it was pretty random and its value didn't fall into any of mapped memory regions.

`spawn_error` pointer was not null and code tried to read its field but this caused segfault instantly. I believe there was no message because for some reason glib did not set it to some value (there was an error!).

Though the basis for that is because mentioned `gio open` was considered to be one argument, it was failing to find such program and hence `g_spawn_sync` returned an error to that.

So this fixes undefined behaviour for that pointer and moreover made both pointers null by default just in case. Not sure why glib does not set the pointer value.

Also I made the `avatar open` use the preference string as a command template rather than an argv argument.

### CODE REVIEW

File: src/xmpp/avatar.c
Comment (sjaeckel) [inline src/xmpp/avatar.c:354]:
```suggestion
        gchar* argv[] = { prefs_get_string(PREF_AVATAR_CMD), filename->str, NULL };
        if (argv[0] == NULL)
            cons_show_error("No default `avatar open` command found in executables preferences.");
        else if (!call_external(argv, NULL, NULL)) {
            cons_show_error("Unable to display avatar: check the logs for more information.");
```

I like the extended check but TBH I'd prefer to keep the old style and simply do the check as suggested

**Review by sjaeckel — COMMENTED** [review]
:+1: 

good catch

**Comment by sjaeckel** [comment]
> `spawn_error` pointer was not null and code tried to read its field but this caused segfault instantly. I believe there was no message because for some reason glib did not set it to some value (there was an error!).

That's strange and sounds like a bug inside glib, since [the documentation clearly states](https://docs.gtk.org/glib/func.spawn_sync.html):

> `error` [...]
> The argument will be set to NULL by the function if there are no errors.
> In case of error, the argument will be set and the caller will take ownership of the data, and be responsible for freeing it."

File: src/xmpp/avatar.c
Comment (nandesu-utils) [reply src/xmpp/avatar.c:354]: I mean this one as will render it impossible to specify "gio open %p" because making it first argument in an array makes `gio open` an entire one argument. It doesn't push `open` as second one. `format_call_external_argv` on the other hand handles spaces.

I think that was the reason why the issue owner had the error initially which then coincidentally resulted in a crash due to another reason (undefined behavior). `gio open` was just not found as a program.

**Comment by nandesu-utils** [comment]
> shows that glib simply returns an error w/o setting `error` if `argv[0]` is `NULL`, therefor the check before calling `call_external()` should be sufficient to fix this issue.

The exit status was not zero (or spawn result, sorry I don't remember). The code was like `255` or something.

Hence this is not about `argv[0]` being null. See, `gio open` was as the first argument and `g_spawn_sync` failed to find the executable file (because there was no executable in the `PATH` named `gio open`, it was not just `gio`, see the batch comment). I think this is a bug, the exit code (wait code) is not zero due to inability to find the executable file.

Just tested out with setting the executable to plain `chromium` and it went fine without crashing.

File: src/xmpp/avatar.c
Comment (nandesu-utils) [reply src/xmpp/avatar.c:354]: Though the failed code style check made me cry a little bit 😢

**Comment by nandesu-utils** [comment]
Okay I've spent some time with unfamiliar glib and read through many a page.

In ac96a2d8781fbe87a3ecbf28619825b877cf0316 I refactored the `call_external` inner body which now corresponds to what I found.

Also note: **this may indicate a bug in glib**. I am not sure, but in fact `g_spawn_sync` **does not** set the error pointer to null when there is no error accordingly to [this](https://docs.gtk.org/glib/func.spawn_sync.html). Moreover, it is completely messed up when there is an error but the error pointer is not null initially, before it is set somewhere in the glib code. Can somebody reproduce? I'll be thankful if I'm wrong. I've looked through glib code but didn't understand much of it right away.

So far, when there is a spawn error, we print it. Spawn error means that we could not just spawn a task for some reason. Otherwise we examine whether the process which was spawned has exited abnormally i.e. returned a non-zero exit code. If it has, we just print an exit error.

Call external function returns true only and only if process was spawned successfully and exited successfully as well, obviously.

In before there were unnecessary checks for `error->message` and basically it would not print an error if process exited with a non-zero code. It is very unlikely that a non-zero code may indicate success so it is now printed as well.

File: src/xmpp/avatar.c
Comment (sjaeckel) [reply src/xmpp/avatar.c]: > I mean this one as will render it impossible to specify "gio open %p" because making it first argument in an array makes gio open an entire one argument. It doesn't push open as second one. format_call_external_argv on the other hand handles spaces.

Oh alright, that makes a lot of sense!

**Comment by sjaeckel** [comment]
> Also note: **this may indicate a bug in glib**

From reading the sources this is a bug. glib makes a promise that it doesn't hold. Will you open a bug with glib or should I? Fixing it isn't that straight-forward, since one would have to touch several places, so it should IMO first be discussed with the glib team how it should be fixed... Whether to fix the code or the documentation ;)

Could you please squash those commits together into one commit with a good commit message? ... and then force-push :)

File: src/common.c
Comment (sjaeckel) [inline src/common.c:469]: you're potentially de-referencing a `NULL`-pointer here
```suggestion
        log_error("could not spawn '%s' with error '%s'", cmd, spawn_error ? spawn_error->message : "Unknown, spawn_error is NULL");
```

File: src/common.c
Comment (sjaeckel) [inline src/common.c:483]: `g_error_free()` has to be called in every case, `exit_error` will always be set when calling `g_spawn_check_wait_status()`

```suggestion
            g_free(cmd);
        }
        g_error_free(exit_error);
    }
```

**Review by sjaeckel — CHANGES_REQUESTED** [review]

File: src/common.c
Comment (nandesu-utils) [reply src/common.c:480]: Is that just a clarity thing or?

According to [this](https://docs.gtk.org/glib/func.spawn_check_wait_status.html#return-value) the error won't be set when process exited successfully thus moving `g_error_free(exit_error)` out of `if (!is_successful) { ... }` block is unnecessary (there is a null check within g_error_free anyway so that is okay).

File: src/common.c
Comment (nandesu-utils) [reply src/common.c:469]: If that has to deal with that `g_spawn_sync` has weird behavior about setting errors then OK. Otherwise `is_successful == FALSE` leads to `spawn_error` to be set to some value hence `spawn_error` is not null.

File: src/common.c
Comment (sjaeckel) [reply src/common.c:480]: You're right (and the documentation as well :) ), I missed the happy path when looking at the implementation and thought it will always set the error

It's fine as is.

File: src/common.c
Comment (sjaeckel) [reply src/common.c:469]: Yes, that's to deal with the weird behavior of `g_spawn_sync`.

Comment (nandesu-utils) [reply src/common.c:469]: I believe we better submit a bug report for glib because the problem is that `spawn_error` stays uninitialized no matter what, or is set to a valid error otherwise. I haven't observed anything else except with a case when `spawn_error` is uninitialized in the first place.

Maybe we'll better add a custom error log message indicating there's something weird with glib instead of just doing `spawn_error ? spawn_error->message : "Unknown, spawn_error is NULL"`?

**Comment by nandesu-utils** [comment]
> From reading the sources this is a bug. glib makes a promise that it doesn't hold. Will you open a bug with glib or should I? Fixing it isn't that straight-forward, since one would have to touch several places, so it should IMO first be discussed with the glib team how it should be fixed... Whether to fix the code or the documentation ;)

I'd ask you to file a report because I'm sure I'm going to miss some information.

> Could you please squash those commits together into one commit with a good commit message? ... and then force-push :)

Yep, of course, but beforehand I'd like to resolve whether we should actually not use a ternary operator there (see comments).

File: src/common.c
Comment (sjaeckel) [reply src/common.c:469]: IMO that'd be already a "custom error log message" since that will only pop up once this exact issue in glib is hit.

@jubalh you're the official maintainer, what do you think?

Comment (jubalh) [reply src/common.c:469]: I would prefer `"Unknown, spawn_error is NULL"` compared to a `its glibs fault!`. I think this error is clear enough, telling us whats going on and easy to grep for in code.

**Comment by sjaeckel** [comment]
> I'd ask you to file a report [...]

https://gitlab.gnome.org/GNOME/glib/-/issues/2732

> [...] because I'm sure I'm going to miss some information.

I hope I didn't miss any ;)

File: src/common.c
Comment (jubalh) [inline src/common.c:450]: this is a little too generic for my taste. If you don't like `spawn_result` what do you think about something like `spawn_succesful`?

**Comment by jubalh** [comment]
@nandesu-utils when you squash/clean up the commits. Could you please add a note about the glib thing to the corresponding commit? Including the link to https://gitlab.gnome.org/GNOME/glib/-/issues/2732

**Comment by sjaeckel** [comment]
GLib issues is closed! Lesson learned: 

> if you are passing a GError as an out argument, you’re supposed to set it to NULL: https://docs.gtk.org/glib/error-reporting.html

File: src/common.c
Comment (nandesu-utils) [reply src/common.c:450]: > this is a little too generic for my taste. If you don't like `spawn_result` what do you think about something like `spawn_succesful`?

I reused the same variable for whether the spawn was successful and also whether the exit status was OK or not.

Otherwise we'd introduce two boolean variables `spawn_successful` and `exited_successfully` respectively with an `&&` at the return statement for clarity.

Or just rework the code further making it explicitly return in places like:

```c
if (spawn_successful) {
    if (g_check_spawn_wait_status(...)) {
        return TRUE;
    } else {
        ...;
        return FALSE;
    }
} else {
    ...;
    return FALSE;
}
```

I don't know honestly. This seems like over-thinking to me as well.

**Comment by nandesu-utils** [comment]
> GLib issue is closed! Lesson learned:
> 
> > if you are passing a GError as an out argument, you’re supposed to set it to NULL: https://docs.gtk.org/glib/error-reporting.html

When-you-read-the-docs-but-actually-they-are-quite-sparse-and-misleading-because-they-are-inconsistent-and-barely-contain-some-information-on-details™️

I wonder if there is something that neglects setting error pointers to null in the codebase at the moment? I am quite unfamiliar with it.

File: src/common.c
Comment (jubalh) [reply src/common.c:450]: yeah true its not a big deal. you can also leave it.

**Comment by jubalh** [comment]
> Also I wonder if there is something that neglects setting error pointers to null in the codebase at the moment? I am quite unfamiliar with it.

After reading the comments here I also thought about this.
But a quick `ack --cc GError` brought:
```
src/command/cmd_funcs.c
7228:        GError* error = NULL;
7237:        GError* error = NULL;

src/config/cafile.c
74:    GError* glib_error = NULL;

src/config/preferences.c
88:    GError* err = NULL;

src/omemo/omemo.c
148:    GError* error = NULL;
606:    GError* error = NULL;
622:    GError* error = NULL;
638:    GError* error = NULL;
648:    GError* error = NULL;
1498:    GError* error = NULL;

src/pgp/gpg.c
190:        GError* gerr = NULL;

src/pgp/ox.c
456:    GError* error = NULL;

src/plugins/plugins.c
181:    GError* error = NULL;

src/tools/editor.c
58:    GError* glib_error = NULL;

src/ui/notifier.c
218:        GError* error = NULL;

src/ui/tray.c
85:    GError* err = NULL;

src/xmpp/avatar.c
105:    GError* err = NULL;
331:    GError* err = NULL;

src/main.c
92:    GError* error = NULL;

src/common.c
88:    GError* error = NULL;
448:    GError *spawn_error, *status_error;
```

So funny enough only this one time we don't do this :)

File: src/common.c
Comment (sjaeckel) [reply src/common.c:469]: This is missing

**Review by sjaeckel — CHANGES_REQUESTED** [review]
LGTM after the re-opened discussion has been taken care of

**Comment by nandesu-utils** [comment]
I think that change request is broken.

**Comment by jubalh** [comment]
> also now code uses `g_spawn_check_wait_status` which
`g_spawn_check_exit_status` has been aliased to.

g_spawn_check_wait_status is only available since glib 2.70. So far we require only 2.62.
Last I checked most distros still had a version < 2.70 but not sure about the current situation.

**Comment by nandesu-utils** [comment]
> > also now code uses `g_spawn_check_wait_status` which
> > `g_spawn_check_exit_status` has been aliased to.
> 
> g_spawn_check_wait_status is only available since glib 2.70. So far we require only 2.62. Last I checked most distros still had a version < 2.70 but not sure about the current situation.

OK! Will change back.

**Comment by jubalh** [comment]
@nandesu-utils ping :)

File: src/common.c
Comment (jubalh) [reply src/common.c:469]: I'll do this now

**Review by jubalh — APPROVED** [review]
i squashed the last two commits together and also changed the error message according to our previous  review. all good now!

**Comment by jubalh** [comment]
@nandesu-utils thanks for your contribution!

**Pull request merged**
